robot/impl/robot_reconfigure_weak_dependencies_test.go

```diff
@@ -285,4 +291,363 @@ func TestUpdateWeakDependents(t *testing.T) {
 	test.That(t, weak1.resources, test.ShouldHaveLength, 2)
 	test.That(t, weak1.resources, test.ShouldContainKey, base1Name)
 	test.That(t, weak1.resources, test.ShouldContainKey, base2Name)
+	test.That(t, weak1.reconfigCount, test.ShouldEqual, 3)
+}
+
+func TestWeakDependentsExplicitDependency(t *testing.T) {
+	// This test tests the behavior of weak dependents that have an explicit dependency.
+	// The robot has 3 components:
+	//   * base1 and base2: two fake base components.
+	//   * weak1: a weak component that depends on all components and also has an explicit dependency on base1.
+	logger := logging.NewTestLogger(t)
+
+	// Register a `Resource` that generates weak dependencies. Specifically instance of
+	// this resource will depend on every `component` resource.
+	weakAPI := resource.NewAPI(uuid.NewString(), "component", "weak")
+	weakModel := resource.DefaultModelFamily.WithModel(utils.RandomAlphaString(5))
+	weak1Name := resource.NewName(weakAPI, "weak1")
+	resource.Register(
+		weakAPI,
+		weakModel,
+		resource.Registration[*someTypeWithWeakAndStrongDeps, *someTypeWithWeakAndStrongDepsConfig]{
+			Constructor: func(
+				ctx context.Context,
+				deps resource.Dependencies,
+				conf resource.Config,
+				logger logging.Logger,
+			) (*someTypeWithWeakAndStrongDeps, error) {
+				return &someTypeWithWeakAndStrongDeps{
+					Named:     conf.ResourceName().AsNamed(),
+					resources: deps,
+				}, nil
+			},
+			WeakDependencies: []resource.Matcher{resource.TypeMatcher{Type: resource.APITypeComponentName}},
+		})
+	defer resource.Deregister(weakAPI, weakModel)
+
+	// Start robot with two components and one resource with weak dependencies
+	// (that also has an explicit dependency on one of the components).
+	// The following scenario is expected:
+	// 1) The two bases will configure first in parallel (which will each bump the logical clock).
+	// 2) The completeConfig loop will check to see if there is a need to updateWeakDependents by
```

Comment:
I mentioned offline that I was worried one of these 5 `updateWeakDependents` calls might be unnecessary. Now, I can't convince myself that we can remove any of them 😆 . In any case, let's not worry about the other `updateWeakDependents` callsites in this PR.

Reply:
This mostly makes sense to me, thanks! Maybe I asked this question in the wrong test. I think the case I've asked you to describe is when `base1` strongly depends on `weak1`. And in this test was just have `weak1` strongly _and_ weakly depending on `base1`.

> Potentially at the end of each level in completeConfig

That reads to me as:
* First we reconfigure base1+base2
* Then we reconfigure weak1

Which sounds fine because we want to update weak1 because things it depended on changed.

But then the sub-bullet says:
> a. This one is new and only happens when "any resources in the level needs reconfiguration and depends on weak dependents, and the resource graph has updated since the last time weak dependents were updated"

Editorializing to make sure we're on the same page:
> any resources in the level needs reconfiguration (e.g: `base1`) and depends on weak dependents (e.g: `weak1`), and the resource graph has updated since the last time weak dependents (e.g: `base1`) were updated

That to me reads as we're reconfiguring the weak dependents because something else depends on them. Normally we want to reconfigure "leaf nodes" first, then work our way up (or down?) to the things that depended on those leaf nodes. But my interpretation of this algorithm is that we're intentionally updating things in the opposite order that we normally do?

Reply:
> Potentially at the end of each level in completeConfig
is inaccurate based on the code - I'm pretty sure `updateWeakDependents` is being run in the beginning of the loop, and that is the intention because of the case Dan is describing

Reply:
> I'm pretty sure updateWeakDependents is being run in the beginning of the loop, and that is the intention because of the case Dan is describing

You're referencing [this call to updateWeakDependents](https://github.com/viamrobotics/rdk/blob/baf6c50963ec05ba848e4891deb2b938e1ba25a6/robot/impl/resource_manager.go?plain=1#L657)?